### ISSUE TITLE
Cors not working with graphql

### ISSUE THREAD

**Issue by bouhendfaycal** [strapi/strapi#8173]
**Describe the bug**
i am sending a mutation with a custom header that I added in the cors middleware config but it still gives error
`Access to fetch at 'http://localhost:1337/graphql' from origin 'http://localhost:3000' has been blocked by CORS policy: Request header field captcha-response is not allowed by Access-Control-Allow-Headers in preflight response.`

**Steps to reproduce the behavior**

1. add custom header to mutation
2. add middleware config
3. try the mutation

middleware.js
```
module.exports = ({ env }) => ({
  settings: {
    cors: {
      enabled: true,
      origin: ['http://localhost:3000'],
      headers: [
        'Content-Type',
        'Authorization',
        'X-Frame-Options',
        'Captcha-Response',
      ],
    },
  },
});
```

**System**

- Node.js version: node:10.16.0-alpine
- NPM version: 6.9.0
- Strapi version: 3.1.6
- Database: mongodb
- Operating system: docker

**Comment by derrickmehaffy** [comment]
Can you please add the system information to your bug report:

```
**System**

- Node.js version: <!-- Please ensure you are using the Node LTS version (v12) -->
- NPM version:
- Strapi version: <!-- Beta and Alpha versions are no longer supported -->
- Database:
- Operating system:
```

**Comment by bouhendfaycal** [comment]
@derrickmehaffy done

**Comment by derrickmehaffy** [comment]
Ty!

**Comment by francois2metz** [comment]
I tried to reproduce the issue without success. Could you share the code that trigger the request?

**Comment by derrickmehaffy** [comment]
I am also unable to reproduce this

**Comment by derrickmehaffy** [comment]
As user has provided no reproduction steps, marking as closed.

**Comment by eikaramba** [comment]
i have the same problem. basically just setup cors via middleware
```
module.exports = [
  'strapi::errors',
  'strapi::security',
  {
    name: "strapi::cors",
    config: {
      enabled: true,
      origin: ['http://localhost:3000','http://localhost:8000'],
      maxAge: 31536000,
      credentials: true,
      methods: [
        "GET",
        "POST",
        "PUT",
        "PATCH",
        "DELETE",
        "OPTIONS",
        "HEAD"
      ],
      headers: [
        "Content-Type",
        "Authorization",
        "X-Frame-Options",
        "Origin",
        "Access-Control-Allow-Headers",
        "Access-Control-Allow-Origin"
      ]
    },
  },
  'strapi::poweredBy',
  'strapi::logger',
  'strapi::query',
  {
    name: "strapi::body",
    config: {
      enabled: true,
      multipart: true,
      includeUnparsed: true
    }
  },
  'strapi::favicon',
  'strapi::public',
];
```

now with the normal api under /api/xxx it works and the cors headers are set.

however if i try to request /graphql cors is not applied and the origin header always returns `*`

**Comment by derrickmehaffy** [comment]
`strapi::cors` doesn't exist, CORs is included in the security middleware

**Comment by eikaramba** [comment]
@derrickmehaffy is it tough? the documentation sais otherwise https://docs.strapi.io/developer-docs/latest/setup-deployment-guides/configurations/required/middlewares.html#internal-middlewares-configuration-reference

also if i remove this cors function, the normal requests to /api/xxx are not using that cors configuration anymore. so i would argue that code is correct, only that graphql endpoint is not making use of it. ;)

**Comment by eikaramba** [comment]
i might found the culprit. it seems graphql is set to false by the graphl plugins config file see here https://github.com/strapi/strapi/blob/master/packages/plugins/graphql/server/bootstrap.js#L55

**Comment by eikaramba** [comment]
after fidling and trying to extend the graphql plugin (did not work), i found the solution:

add this line to https://github.com/strapi/strapi/blob/master/packages/plugins/graphql/server/bootstrap.js#L98 will actually make apolloServer use the provided origin definition (which still uses a config provided from overwritting the graphql plugin)
```js
server.applyMiddleware({ app:strapi.server.app, cors: config('apolloServer').cors});
```

**summary:**
- specifying the cors policy within the strapi::cors middleware is not affecting apolloServer in V4
- overwriting the apolloServer config is not working as well, see https://stackoverflow.com/a/54589681/2644039. i can confirm this as having a /src/plugins/graphql/strapi-server.js file with
```js
module.exports = (plugin) => {
    plugin.config.apolloServer.cors = {
      origin: (origin) => {
        const whitelist = [
            "http://localhost:3000",
        ];
        const found = whitelist.indexOf(origin.request.header.origin);
        if (found !== -1) {
            return found
        } else {
          return "null";
        }
    },
      credentials:true
    };
    return plugin;
  };
```
is not actually working. the middleware NEEDS to be registered.


From my point of view there are 2 things to do here:
- add middleware registration to bootstrap.js
- somehow read cors values from strapi::cors middleware for apolloserver instead of needing to overwrite graphql-server.js file.

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy is it tough? the documentation says otherwise https://docs.strapi.io/developer-docs/latest/setup-deployment-guides/configurations/required/middlewares.html#internal-middlewares-configuration-reference also the code shows it's there: https://github.com/strapi/strapi/blob/master/packages/core/strapi/lib/middlewares/cors.js
> 
> also if i remove this cors function, the normal requests to /api/xxx are not using that cors configuration anymore. so i would argue that code is correct, only that graphql endpoint is not making use of it. ;)

I'm sorry your right, CORs isn't in the security middleware, we switched to using koa-helment and I thought it was in that one.

**Comment by eikaramba** [comment]
ok let me know if you cannot reproduce it or something like that. i have my client on 3000 and strapi on 1337. and use fetch with credentials:true - at least FF(but chrome should too) preventing CORS request where allow credentials is true and origin is `*` (because this is super unsecure)

**Comment by sebastian-ludwig** [comment]
I have got the same setup and issue here: Preflight is using the correct Settings, but the POST not.


![Bildschirmfoto 2022-01-17 um 23 20 40](https://user-images.githubusercontent.com/4415200/149843164-8afbec9d-5605-49c5-acff-be33be195165.png)
![Bildschirmfoto 2022-01-17 um 23 21 51](https://user-images.githubusercontent.com/4415200/149843173-290999e7-5aa9-4e8c-9674-08e93341c087.png)

```
server.getMiddleware({ path:path,cors: config('apolloServer').cors }),
```
worked for me as well 👍

**Comment by eikaramba** [comment]
i have now switched to Directus as my go to backend framework given that strapi v4 is not usable right now with graphql endpoint and CORS. hopefully at some point the strapi team decides to fix this here, as it is quite a basic requirement.

**Comment by Nima155** [comment]
I am also facing this issue. This is really such a deal-breaker and I hope that you guys can fix it soon.

**Comment by eikaramba** [comment]
@derrickmehaffy nope, that is for v4. why do you think it should not be an issue for v4?

i haven't used strapi because of this in the last 3 months, but i can assure you it was definitely not working in v4, see the comments at the start of this issue.

**Comment by derrickmehaffy** [comment]
> @derrickmehaffy nope, that is for v4. why do you think it should not be an issue for v4?
> 
> i haven't used strapi because of this in the last 3 months, but i can assure you it was definitely not working in v4, see the comments at the start of this issue.

I was test on v4.1.11 and couldn't reproduce, the original issue was opened on Oct 3rd, 2020 long before v4 was around.

If users still hit this issue it would probably be better to open a new issue so we can get reproduction steps.

**Comment by inDivyanshu** [comment]
I am facing the same issue, with any plugin I develop. we are able to work with only GET APIs , all the others throw preflight issue
